### PR TITLE
Add countdown command

### PR DESCRIPTION
**Pull request by DannyLeWasTaken**

### CODE REVIEW

**Comment by DannyLeWasTaken** [comment]
@notevanrutten Code is ready for you to review

File: src/message-manager.mts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    // Channel of countdown
+    [channelId: string]: {
+        messageId: string;
+        events: {
+            // Events of countdown
+            [eventName: string]: {
+                eventDate: Date; // Date of countdown
+                eventLink: string; // Any link if given
+            };
+        };
```
Comment (Nomalah) [inline src/message-manager.mts:27]: Alternatively, you could do:
```suggestion
        events: CountdownMessageInput[];
```

File: src/message-manager.mts [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    };
+};
+
+// Deserialize messages
+const messageDictionary: Countdowns = fs.existsSync("./messages.json")
```
Comment (Nomalah) [inline src/message-manager.mts:32]: Every time this is imported, you're going to read that file synchronously over and over + style guide

File: src/message-manager.mts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+                timeLeft = Math.round(deltaMonths) + " month(s)";
+            } else if (deltaWeeks > 2) {
+                timeLeft = Math.round(deltaWeeks) + " week(s)";
+            } else if (deltaDays > 3) {
+                timeLeft = Math.round(deltaDays) + " day(s)";
+            } else {
+                timeLeft = deltaHours + " hour(s)";
+            }
+
+            fields.push({ name: `${countdownName}`, value: `[${event_locale}](${countdown.eventLink})\nTime remaining: ${timeLeft}` });
```
Comment (Nomalah) [inline src/message-manager.mts:98]: Style guide

File: src/message-manager.mts [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            return;
+        } else if (!terminate_on_message_destruction) {
+            message = await channel.send({ embeds: [embedded] });
+            messageDictionary[channel_id].messageId = message.id;
+            updateMessageDictionary();
+        }
+    }
+
+    // If the message is older than 24 hours, delete and make a new one
+    if ((message && now.getTime() - message.createdAt.getTime() >= 1000 * 60 * 60 * 12) || force_new_message) {
```
Comment (Nomalah) [inline src/message-manager.mts:123]: Just clean up this logic (it's also 12 hours)

**Review by Nomalah — CHANGES_REQUESTED** [review]
Mainly style guide issues, clean up the code and document it better, ex:
```ts
// Adds a countdown
export const addCountdown = (client: Client, channelId: string, messageInput: CountdownMessageInput) => {
```
^ Duh

File: src/commands/countdownAdd.mts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        const options = interaction.options;
+        const dateString = options.get("date")?.value?.toString();
+        const eventName = options.get("name")?.value?.toString();
+        const eventLink = options.get("url")?.value?.toString();
+        if (dateString === undefined && eventName === undefined) {
+            // This shouldn't be possible as we set those two a requirements
+            return;
+        }
+        console.log(dateString);
+        const match = dateString?.match(/\b(\d{4})\/(0[1-9]|1[0-2])\/(0[1-9]|[12][0-9]|3[01])\b/g);
```
[the remark this answers is not in the document]
Comment (DannyLeWasTaken) [reply src/commands/countdownAdd.mts]: I literally asked ChatGPT

File: src/commands/countdown.mts [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+                if (match === null) {
+                    await interaction.reply({
+                        content: "Invalid date format. Please use YYYY/MM/DD format",
+                        ephemeral: true,
+                    });
+                    return;
+                }
+                const [year, month, day] = match![0].split("/");
+                const date = new Date(Number(year), Number(month) - 1, Number(day));
+                if (date.getTime() - new Date().getTime() > 0) {
```
[the remark this answers is not in the document]
Comment (DannyLeWasTaken) [reply src/commands/countdown.mts:64]: Mhm mb

File: src/commands/countdown.mts [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                    });
+                    await interaction.reply({ content: "Countdown added", ephemeral: true });
+                } else {
+                    // All countdowns must be in the future
+                    await interaction.reply({ content: "Date specified should be in the future", ephemeral: true });
+                }
+                break;
+            }
+            case "remove": {
+                const eventName = options.get("name")?.value?.toString();
```
[the remark this answers is not in the document]
Comment (DannyLeWasTaken) [reply src/commands/countdown.mts:83]: Casting to it as a string

File: src/message-manager.mts (hunk 2 above)
Comment (DannyLeWasTaken) [reply src/message-manager.mts:32]: Singleton?

File: package.json [hunk 8]
```diff
@@ -2,14 +2,18 @@
...
     "dev": "tsc --watch",
     "build": "tsc --strict",
+    "run":  "node dist/index.mjs",
+    "deploy-commands": "node dist/deploy-commands.mjs",
     "lint": "eslint . --ext .vue,.js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --ignore-path .gitignore --fix",
     "type-check": "tsc --strict --noEmit",
     "lint-check": "eslint . --ext .vue,.js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --ignore-path .gitignore"
   },
   "dependencies": {
+    "@types/node-cron": "^3.0.8",
```
Comment (Nomalah) [inline package.json:12]: Types should be a dev dependency

File: src/commands/countdown.mts [hunk 9]
```diff
@@ -1,12 +1,102 @@
...
+        switch (options.getSubcommand()) {
+            case "add": {
+                const date_string = options.get("date")?.value?.toString();
+                const event_name = options.get("name")?.value?.toString();
+                const event_link = options.get("url")?.value?.toString();
+                if (date_string === undefined && event_name === undefined) {
+                    // This shouldn't be possible as we set those two a requirements
+                    return;
+                }
+                // https://developers.redhat.com/articles/2022/10/13/advanced-regex-capture-groups-lookaheads-and-lookbehinds
```
Comment (Nomalah) [inline src/commands/countdown.mts:53]: Just explain what the regex is, and what it's trying to match.

File: src/commands/countdown.mts [hunk 10]
```diff
@@ -1,12 +1,102 @@
...
+            }
+            case "remove": {
+                const event_name = options.get("name")?.value?.toString();
+                if (event_name === undefined) {
+                    // This shouldn't be possible as we set those two a requirements
+                    return;
+                }
+                deleteCountdown(interaction.client, interaction.channelId, event_name);
+                updateMessage(interaction.client, interaction.channelId, true, true, null).then(() => {
+                    // Update message every 5 minute. We cannot assume a cron schedule
```
Comment (Nomalah) [inline src/commands/countdown.mts:92]: But *why* can't we assume a cron schedule already exists

File: src/countdown-manager.mts [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+    }
+    message_dictionary[channelId].events[messageInput.event_name] = {
+        event_date: messageInput.event_date,
+        event_link: messageInput.event_link === null ? "https://www.youtube.com/watch?v=dQw4w9WgXcQ" : messageInput.event_link,
+    };
+    updateMessageDictionary();
+};
+
+// Removes a countdown
+export const deleteCountdown = (client: Client, channelId: string, eventName: string) => {
```
Comment (Nomalah) [inline src/countdown-manager.mts:153]: For named functions, favour functions over lambdas. + style guide

File: src/countdown-manager.mts [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        message = await channel.send({ embeds: [embedded] });
+        message_dictionary[channel_id].message_id = message.id;
+        updateMessageDictionary();
+    }
+}
+
+// Adds a countdown
+export const addCountdown = (client: Client, channelId: string, messageInput: CountdownMessageInput) => {
+    if (!message_dictionary[channelId]) {
+        message_dictionary[channelId] = {
```
Comment (Nomalah) [inline src/countdown-manager.mts:140]: Style guide

**Review by Nomalah — CHANGES_REQUESTED** [review]
Still style guide issues

File: src/countdown-manager.mts [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+export function addCountdown(client: Client, channelId: string, messageInput: CountdownMessageInput) {
+    if (!message_dictionary[channelId]) {
+        message_dictionary[channelId] = {
+            message_id: "",
+            events: {},
+        };
+    }
+    message_dictionary[channelId].events[messageInput.event_name] = {
+        event_date: messageInput.event_date,
+        event_link: messageInput.event_link === null ? "https://www.youtube.com/watch?v=dQw4w9WgXcQ" : messageInput.event_link,
```
Comment (Nomalah) [inline src/countdown-manager.mts:147]: I knew that typing looked a little off...

File: src/countdown-manager.mts [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+    };
+    updateMessageDictionary();
+}
+
+// Removes a countdown
+export function deleteCountdown(client: Client, channelId: string, eventName: string) {
```
Comment (Nomalah) [inline src/countdown-manager.mts:153]: Style guide + remove comment

**Review by Nomalah — CHANGES_REQUESTED** [review]
You almost got me there...

File: src/countdown-manager.mts (hunk 14 above)
Comment (DannyLeWasTaken) [reply src/countdown-manager.mts:153]: Fixed linting and linting only

**Review by Nomalah — APPROVED** [review]
LGTM

**Pull request merged**